lib/Dialect/Moore/Transforms/SimplifyAssigns.cpp

```diff
@@ -0,0 +1,83 @@
+//===- DeleteLocalVar.cpp - Delete local temporary variables --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the SimplifyAssigns pass.
+// It's used to disassemble the moore.concat_ref. Which is tricky to lower
+// directly. For example, disassemble "{a, b} = c" onto "a = c[7:3]"
+// and "b = c[2:0]".
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Moore/MooreOps.h"
+#include "circt/Dialect/Moore/MoorePasses.h"
+
+namespace circt {
+namespace moore {
+#define GEN_PASS_DEF_SIMPLIFYASSIGNS
+#include "circt/Dialect/Moore/MoorePasses.h.inc"
+} // namespace moore
+} // namespace circt
+
+using namespace circt;
+using namespace moore;
+
+namespace {
+struct SimplifyAssignsPass
+    : public circt::moore::impl::SimplifyAssignsBase<SimplifyAssignsPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+std::unique_ptr<mlir::Pass> circt::moore::createSimplifyAssignsPass() {
+  return std::make_unique<SimplifyAssignsPass>();
+}
+
+void SimplifyAssignsPass::runOnOperation() {
+  getOperation()->walk([&](ConcatRefOp concatRefOp) {
+    mlir::OpBuilder builder(&getContext());
+
+    // moore.concat_ref only as the LHS of assignments.
+    if (auto *user = *concatRefOp->getUsers().begin()) {
```

Comment:
I was thoughtless. You're absolutely right! The `moore.instance` also uses `moore.concat_ref`, such as inout ports.

Reply:
I can promise every `moore.concat_ref` only has a single user, because we will always create a new `moore.concat_ref` when we need it. This can be checked in the `Expressions.cpp`-- `Value Context::convertLvalueExpression(const slang::ast::Expression &expr){...}`

Reply:
No, `concat_ref` is marked as `Pure` so CSE would freely eliminate duplicated values (so multiple users). If we want to keep a single user as an invariant we need to eliminate `Pure` and add a verifier to verify that but I don't think it's reasonable.

Reply:
Learned new knowledge.